src/engine/controls/bpmcontrol.cpp

```diff
@@ -725,6 +726,142 @@ double BpmControl::getNearestPositionInPhase(
+    }
+
+    TrackPointer otherTrack = pOtherEngineBuffer->getLoadedTrack();
+    BeatsPointer otherBeats = otherTrack ? otherTrack->getBeats() : BeatsPointer();
+
+    // If either track does not have beats, then we can't adjust the phase.
+    if (!otherBeats) {
+        return dThisPosition;
+    }
+
+    double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
+
+    if (!BpmControl::getBeatContext(otherBeats, dOtherPosition,
+                  &dOtherPrevBeat, &dOtherNextBeat, &dOtherBeatLength, &dOtherBeatFraction)) {
+        return dThisPosition;
+    }
+
+    double dThisSampleRate = m_pBeats->getSampleRate();
+    double dThisRateRatio = m_pRateRatio->get();
+
+    // Seek our next beat to the other next beat
+    // This is the only thing we can do if the track has different BPM,
+    // playing the next beat together.
+    double thisDivSec = (dThisNextBeat - dThisPosition)
+            / dThisSampleRate / dThisRateRatio;
+
+    // dOtherBeatFraction =+ m_dUserOffset;
+    double otherDivSec = (1 - dOtherBeatFraction) * dOtherBeatLength
+            / otherBeats->getSampleRate() / pOtherEngineBuffer->getRateRatio();
+
+    double dNewPlaypos = dThisPosition +
+            (thisDivSec - otherDivSec) *
+            dThisSampleRate * dThisRateRatio;
+
+    if (respectLoops) {
+        // We might be seeking outside the loop.
+        const bool loop_enabled = m_pLoopEnabled->toBool();
+        const double loop_start_position = m_pLoopStartPosition->get();
+        const double loop_end_position = m_pLoopEndPosition->get();
+
+        // Cases for sanity:
+        //
+        // CASE 1
+        // Two identical 1-beat loops, out of phase by X samples.
+        // Other deck is at its loop start.
+        // This deck is half way through. We want to jump forward X samples to the loop end point.
+        //
+        // Two identical 1-beat loop, out of phase by X samples.
+        // Other deck is
+
+        // If sync target is 50% through the beat,
+        // If we are at the loop end point and hit sync, jump forward X samples.
+
+
+        // TODO(rryan): Revise this with something that keeps a broader number of
+        // cases in sync. This at least prevents breaking out of the loop.
+        if (loop_enabled &&
+                dThisPosition <= loop_end_position) {
+            const double loop_length = loop_end_position - loop_start_position;
+            const double end_delta = dNewPlaypos - loop_end_position;
+
+            // Syncing to after the loop end.
+            if (end_delta > 0 && loop_length > 0.0) {
+                int i = end_delta / loop_length;
+                dNewPlaypos = loop_start_position + end_delta - i * loop_length;
+
+                // Move new position after loop jump into phase as well.
+                // This is a recursive call, called only twice because of
+                // respectLoops = false
+                dNewPlaypos = getNearestPositionInPhase(dNewPlaypos, false, playing);
```

Comment:
do you mean to recurse into getnearestpositioninphase, or should this be getbeatmatchposition?